lib/experimental/html/class-wp-html-tag-find-descriptor.php

```diff
@@ -0,0 +1,178 @@
+	 * a class name this property will be `null`.
+	 *
+	 * @since 6.1.0
+	 * @var string|null
+	 */
+	private $class_pattern;
+
+	/**
+	 * Used to skip matches in case we expect more than one to exist.
+	 * This constraint applies after all other constraints have held.
+	 * For example, to find the first `<div>` tag containing the
+	 * `wp-block` class name set `match_offset = 0`. To find the third
+	 * match, set `match_offset = 2`. If not provided the default indication
+	 * is to find the first match.
+	 *
+	 * @since 6.1.0
+	 * @var int
+	 */
+	public $match_offset = 0;
+
+	/**
+	 * Creates a tag find descriptor given the input parameters specifying
+	 * the intended match, encodes inputs for searching.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param array|string|WP_HTML_Tag_Find_Descriptor $query {
+	 *     Which tag name to find, having which class, etc.
+	 *
+	 *     @type string|null $tag_name     Which tag to find, or `null` for "any tag."
+	 *     @type int|null    $match_offset Find the Nth tag matching all search criteria.
+	 *                                     0 for "first" tag, 2 for "third," etc.
+	 *                                     Defaults to first tag.
+	 *     @type string|null $class_name   Tag must contain this whole class name to match.
+	 * }
+	 * @return WP_HTML_Tag_Find_Descriptor Used by WP_HTML_Processor when scanning HTML.
+	 */
+	public static function parse( $query ) {
+		static $descriptor = null;
+		static $last_query = null;
+
+		if ( $query instanceof WP_HTML_Tag_Find_Descriptor ) {
+			return $query;
+		}
+
+		if ( null === $descriptor ) {
+			$descriptor = new WP_HTML_Tag_Find_Descriptor();
+		}
+
+		if ( $last_query === $query ) {
+			return $descriptor;
+		}
+
+		$last_query = $query;
+
+		// Reset the descriptor, reusing the already-allocated one.
+		$descriptor->tag_name      = null;
+		$descriptor->class_pattern = null;
+		$descriptor->match_offset  = 0;
+
+		if ( is_string( $query ) ) {
+			$descriptor->tag_name = $query;
+			return $descriptor;
+		}
+
+		if ( ! is_array( $query ) ) {
+			return $descriptor;
+		}
+
+		if ( isset( $query['tag_name'] ) && is_string( $query['tag_name'] ) ) {
+			$descriptor->tag_name = strtolower( $query['tag_name'] );
+		}
+
+		if ( isset( $query['match_offset'] ) && is_integer( $query['match_offset'] ) ) {
+			$descriptor->match_offset = $query['match_offset'];
+		}
+
+		if ( isset( $query['class_name'] ) && is_string( $query['class_name'] ) ) {
+			$descriptor->class_pattern = $query['class_name'];
+		}
+
+		return $descriptor;
+	}
+
+	/**
+	 * Checks whether a given tag and its attributes match the search criteria.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param string                    $tag        The name of the tag.
+	 * @param WP_HTML_Attribute_Token[] $attributes The attributes for the tag.
+	 *
+	 * @return boolean
+	 */
+	public function matches( $walker ) {
+		if ( $this->tag_name ) {
+			$length = $walker->tag_name_ends_at - $walker->tag_name_starts_at;
```

Comment:
`tag_name_ends_at` and `tag_name_starts_at` are marked as private in the walker object which triggers errors:

<img width="1184" alt="Screenshot 2022-08-08 at 12 39 34" src="https://user-images.githubusercontent.com/699132/183401163-1938cc07-2218-46d5-b7f8-d2347459cb34.png">

This needs some further refactoring, as I see that the same issue will happen with all other walker fields that need to be accessed in this method.

Reply:
Reverting the last commit https://github.com/WordPress/gutenberg/pull/42485/commits/7413c113b1f3ea5979ea5896aa958a51b9af9aa2 resolves all the issues.

Reply:
@dmsnell, I added a revert commit https://github.com/WordPress/gutenberg/pull/42485/commits/71fdaeaa2bf9142d9c40474bce42574e29f31c0a. Feel free to remove it locally whenever you want to continue working on the started performance enhancements.